### PR TITLE
Add check-inject feature to describe pod injection

### PR DESCRIPTION
**Pull request by hanxiaop**
**Please provide a description of this PR:**
Fix https://github.com/istio/istio/issues/38299

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Looks like a great start, awesome to see this

File: istioctl/cmd/checkinject.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+			} else {
+				rest, err := client.UtilFactory().ToRESTConfig()
+				if err != nil {
+					return err
+				}
+				coreClient, err := corev1client.NewForConfig(rest)
+				if err != nil {
+					return err
+				}
+				sortBy := func(pods []*corev1.Pod) sort.Interface { return podutils.ByLogging(pods) }
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: @howardjohn Thanks for the clarification. I think using static labels is useful, revised.

File: istioctl/cmd/checkinject.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+				}
+				podName = pod.Name
+				podNs = pod.Namespace
+			}
+			whs, err := client.Kube().AdmissionregistrationV1().MutatingWebhookConfigurations().List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			if len(whs.Items) == 0 {
+				return fmt.Errorf("ERROR: no injector is found for sidecar injection")
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: Thanks, I've revised to option 2.

File: istioctl/cmd/checkinject.go [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+		fmt.Fprintf(writer, "No Istio injection hooks present.\n")
+		return nil
+	}
+	w := new(tabwriter.Writer).Init(writer, 0, 8, 1, ' ', 0)
+	fmt.Fprintln(w, "WEBHOOK\tREVISION\tINJECTED\tREASON")
+	injectedTotal := 0
+	for _, ws := range was {
+		injected := "\u2716"
+		if ws.Injected {
+			injected = "\u2714"
```
[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: @howardjohn Using the color package will mess the tabwriter format up. I left with a TODO, and will investigate if the color can be used.

[the remark this answers is not in the document]
Comment (howardjohn) [reply istioctl/cmd/checkinject.go]: Bummer. I did it in https://github.com/howardjohn/kubectl-resources/blob/master/pkg/writer/tablewriter.go but I think had a lot of custom code

[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: @howardjohn Thanks. Just found a package https://github.com/juju/ansiterm which can be used with the color package. The output looks great.

[the remark this answers is not in the document]
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: @howardjohn Thanks for the infomation. I rewrite the tabwriter referring to the one you wrote, and now it works. PTAL.

**Comment by hanxiaop** [comment]
not stale

File: istioctl/cmd/checkinject.go [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+func checkInjectCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "check-inject [<type>/]<name>[.<namespace>]",
+		Short: "Check the injection status or inject-ability of a given resource, explains why it is (or will be) injected or not",
+		Long: `
+Checks associated resources of the given resource, and running webhooks to examine whether the pod can be or will be injected or not.`,
+		Example: `	# Check the injection status of a pod
+	istioctl check-inject details-v1-fcff6c49c-kqnfk.test
+	
+	# Check the injection status of a pod under a deployment
```
Comment (irisdingbj) [inline istioctl/cmd/checkinject.go:47]: will this check all pods for this deployment?

File: istioctl/cmd/checkinject.go [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+var labelPairs string
+
+func checkInjectCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "check-inject [<type>/]<name>[.<namespace>]",
+		Short: "Check the injection status or inject-ability of a given resource, explains why it is (or will be) injected or not",
+		Long: `
+Checks associated resources of the given resource, and running webhooks to examine whether the pod can be or will be injected or not.`,
+		Example: `	# Check the injection status of a pod
+	istioctl check-inject details-v1-fcff6c49c-kqnfk.test
```
Comment (irisdingbj) [inline istioctl/cmd/checkinject.go:45]: for newly added command, it is better to add them under `experimental` at first

File: istioctl/cmd/checkinject.go (hunk 4 above)
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go:47]: @irisdingbj This command runs on a per-pod basis, and only one pod will be checked under a deployment. It utilizes `InferPodInfoFromTypedResource` function like some of the other commands.

File: istioctl/cmd/checkinject.go (hunk 5 above)
Comment (hanxiaop) [reply istioctl/cmd/checkinject.go]: Thanks, fixed.

**Review by GregHanson — APPROVED** [review]
release note looks good

**Review by irisdingbj — APPROVED** [review]
Thanks for this!

**Pull request merged**